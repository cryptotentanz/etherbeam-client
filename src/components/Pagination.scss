@use '../styles/resources/sizes';

.my-pagination {
  justify-content: center;
  //margin: sizes.$space-s 0px !important;
}
