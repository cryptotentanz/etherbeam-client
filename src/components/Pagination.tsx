import React, { FC } from 'react'
import PaginationItem from './PaginationItem'
import List from './List'

interface PaginationProps {
  total: number
  current: number
  setCurrent: (next: number | ((previous: number) => number)) => void
  className?: string
}

const ELLIPSIS_THRESHOLD_PAGES = 7

const Pagination: FC<PaginationProps> = ({ total, current, setCurrent, className }: PaginationProps) => {
  const displayEllipsis = total > ELLIPSIS_THRESHOLD_PAGES
  const pages: FC[] = []

  Array.from({ length: total }).map((_, index) => {
    const page = index + 1
    const isCurrent = page === current
    if (!displayEllipsis || [1, current - 1, current, current + 1, total].includes(page)) {
      pages.push(() => (
        <PaginationItem page={page} action={() => setCurrent(page)} selected={isCurrent} disabled={isCurrent} />
      ))
    } else if (page == current - 2 || page == current + 2) {
      pages.push(() => <PaginationItem page="ellipsis" disabled />)
    }
  })

  return (
    <nav>
      <List horizontal center className={className}>
        <PaginationItem page="first" action={() => setCurrent(1)} disabled={current === 1} />
        <PaginationItem
          page="previous"
          action={() => setCurrent((previous) => previous - 1)}
          disabled={current === 1}
        />
        {pages.map((Page, index) => (
          <Page key={index} />
        ))}
        <PaginationItem
          page="next"
          action={() => setCurrent((previous) => previous + 1)}
          disabled={current === total}
        />
        <PaginationItem page="last" action={() => setCurrent(total)} disabled={current === total} />
      </List>
    </nav>
  )
}

export default Pagination
