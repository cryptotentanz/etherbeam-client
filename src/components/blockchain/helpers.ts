import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13
const UNISWAP_SWAP_URL = 'https://app.uniswap.org/#/swap'

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string | null, decimals: number, options?: AmountPrintOptions | null): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber | null, options?: AmountPrintOptions | null, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (
  amount: BigNumber | null,
  token: Token,
  options?: AmountPrintOptions | null
): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition | null => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}

export const getUniswapTradeUrl = (inputHash: string | null, outputHash: string | null): string => {
  return `${UNISWAP_SWAP_URL}?${inputHash ? `inputCurrency=${inputHash}&` : ''}${
    outputHash ? `outputCurrency=${outputHash}` : ''
  }`
}
