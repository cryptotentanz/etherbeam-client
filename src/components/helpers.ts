import { UserType } from 'libraries/types'
import { faUser, faUserShield, IconDefinition } from '@fortawesome/free-solid-svg-icons'

export const getUserTypeIcon = (type: UserType): IconDefinition => {
  switch (type) {
    case UserType.Administrator:
      return faUserShield
    default:
      return faUser
  }
}
