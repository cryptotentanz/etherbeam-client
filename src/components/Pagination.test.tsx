import React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { clearMocks } from 'tests/helpers'
import Pagination from './Pagination'

describe('<Pagination />', () => {
  const setCurrent = jest.fn()

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(<Pagination total={6} current={5} setCurrent={setCurrent} />)
    })

    it('should render items', () => {
      const items = component.find('PaginationItem')
      expect(items).toHaveLength(10)
      expect(items.at(0).prop('disabled')).toBeFalsy()
      expect(items.at(1).prop('disabled')).toBeFalsy()
      expect(items.at(2).prop('selected')).toBeFalsy()
      expect(items.at(3).prop('selected')).toBeFalsy()
      expect(items.at(4).prop('selected')).toBeFalsy()
      expect(items.at(5).prop('selected')).toBeFalsy()
      expect(items.at(6).prop('selected')).toBeTruthy()
      expect(items.at(7).prop('selected')).toBeFalsy()
      expect(items.at(8).prop('disabled')).toBeFalsy()
      expect(items.at(9).prop('disabled')).toBeFalsy()
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Ellipsed', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)
    })

    it('should render items', () => {
      const items = component.find('PaginationItem')
      expect(items).toHaveLength(11)
      expect(items.at(0).prop('disabled')).toBeFalsy()
      expect(items.at(1).prop('disabled')).toBeFalsy()
      expect(items.at(2).prop('selected')).toBeFalsy()
      expect(items.at(3).prop('disabled')).toBeTruthy()
      expect(items.at(4).prop('selected')).toBeFalsy()
      expect(items.at(5).prop('selected')).toBeTruthy()
      expect(items.at(6).prop('selected')).toBeFalsy()
      expect(items.at(7).prop('disabled')).toBeTruthy()
      expect(items.at(8).prop('selected')).toBeFalsy()
      expect(items.at(9).prop('disabled')).toBeFalsy()
      expect(items.at(10).prop('disabled')).toBeFalsy()
    })
  })

  describe('Class', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(<Pagination total={6} current={5} setCurrent={setCurrent} className="my-class" />)
    })

    it('should render', () => expect(component.find('List').at(0).hasClass('my-class')).toBeTruthy())
  })

  describe('On click', () => {
    describe('First', () => {
      beforeAll(() => {
        clearMocks()
        setCurrent.mockImplementation((newValue) => {
          component.setProps({ current: newValue })
        })

        component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)

        act(() => {
          component.find('PaginationItem').at(0).find('Button').simulate('click')
          component.update()
        })
      })

      it('should set current', () => {
        expect(setCurrent).toBeCalledTimes(1)
        expect(component.prop('current')).toBe(1)
      })

      it('should render items', () => {
        const items = component.find('PaginationItem')
        expect(items).toHaveLength(8)
        expect(items.at(0).prop('disabled')).toBeTruthy()
        expect(items.at(1).prop('disabled')).toBeTruthy()
        expect(items.at(2).prop('selected')).toBeTruthy()
        expect(items.at(3).prop('selected')).toBeFalsy()
        expect(items.at(4).prop('disabled')).toBeTruthy()
        expect(items.at(5).prop('selected')).toBeFalsy()
        expect(items.at(6).prop('disabled')).toBeFalsy()
        expect(items.at(7).prop('disabled')).toBeFalsy()
      })
    })
    describe('Previous', () => {
      beforeAll(() => {
        clearMocks()
        setCurrent.mockImplementation((newValue) => {
          component.setProps({ current: newValue(5) })
        })

        component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)

        act(() => {
          component.find('PaginationItem').at(1).find('Button').simulate('click')
          component.update()
        })
      })

      it('should set current', () => {
        expect(setCurrent).toBeCalledTimes(1)
        expect(component.prop('current')).toBe(4)
      })

      it('should render items', () => {
        const items = component.find('PaginationItem')
        expect(items).toHaveLength(11)
        expect(items.at(0).prop('disabled')).toBeFalsy()
        expect(items.at(1).prop('disabled')).toBeFalsy()
        expect(items.at(2).prop('selected')).toBeFalsy()
        expect(items.at(3).prop('disabled')).toBeTruthy()
        expect(items.at(4).prop('selected')).toBeFalsy()
        expect(items.at(5).prop('selected')).toBeTruthy()
        expect(items.at(6).prop('selected')).toBeFalsy()
        expect(items.at(7).prop('disabled')).toBeTruthy()
        expect(items.at(8).prop('selected')).toBeFalsy()
        expect(items.at(9).prop('disabled')).toBeFalsy()
        expect(items.at(10).prop('disabled')).toBeFalsy()
      })
    })

    describe('Specific', () => {
      beforeAll(() => {
        clearMocks()
        setCurrent.mockImplementation((newValue) => {
          component.setProps({ current: newValue })
        })

        component = mount(<Pagination total={10} current={3} setCurrent={setCurrent} />)

        act(() => {
          component.find('PaginationItem').at(3).find('Button').simulate('click')
          component.update()
        })
      })

      it('should set current', () => {
        expect(setCurrent).toBeCalledTimes(1)
        expect(component.prop('current')).toBe(2)
      })

      it('should render items', () => {
        const items = component.find('PaginationItem')
        expect(items).toHaveLength(9)
        expect(items.at(0).prop('disabled')).toBeFalsy()
        expect(items.at(1).prop('disabled')).toBeFalsy()
        expect(items.at(2).prop('selected')).toBeFalsy()
        expect(items.at(3).prop('selected')).toBeTruthy()
        expect(items.at(4).prop('selected')).toBeFalsy()
        expect(items.at(5).prop('disabled')).toBeTruthy()
        expect(items.at(6).prop('selected')).toBeFalsy()
        expect(items.at(7).prop('disabled')).toBeFalsy()
        expect(items.at(8).prop('disabled')).toBeFalsy()
      })
    })

    describe('Next', () => {
      beforeAll(() => {
        clearMocks()
        setCurrent.mockImplementation((newValue) => {
          component.setProps({ current: newValue(5) })
        })

        component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)

        act(() => {
          component.find('PaginationItem').at(9).find('Button').simulate('click')
          component.update()
        })
      })

      it('should set current', () => {
        expect(setCurrent).toBeCalledTimes(1)
        expect(component.prop('current')).toBe(6)
      })

      it('should render items', () => {
        const items = component.find('PaginationItem')
        expect(items).toHaveLength(11)
        expect(items.at(0).prop('disabled')).toBeFalsy()
        expect(items.at(1).prop('disabled')).toBeFalsy()
        expect(items.at(2).prop('selected')).toBeFalsy()
        expect(items.at(3).prop('disabled')).toBeTruthy()
        expect(items.at(4).prop('selected')).toBeFalsy()
        expect(items.at(5).prop('selected')).toBeTruthy()
        expect(items.at(6).prop('selected')).toBeFalsy()
        expect(items.at(7).prop('disabled')).toBeTruthy()
        expect(items.at(8).prop('selected')).toBeFalsy()
        expect(items.at(9).prop('disabled')).toBeFalsy()
        expect(items.at(10).prop('disabled')).toBeFalsy()
      })
    })

    describe('Last', () => {
      beforeAll(() => {
        clearMocks()
        setCurrent.mockImplementation((newValue) => {
          component.setProps({ current: newValue })
        })

        component = mount(<Pagination total={10} current={5} setCurrent={setCurrent} />)

        act(() => {
          component.find('PaginationItem').at(8).find('Button').simulate('click')
          component.update()
        })
      })

      it('should set current', () => {
        expect(setCurrent).toBeCalledTimes(1)
        expect(component.prop('current')).toBe(10)
      })

      it('should render items', () => {
        const items = component.find('PaginationItem')
        expect(items).toHaveLength(8)
        expect(items.at(0).prop('disabled')).toBeFalsy()
        expect(items.at(1).prop('disabled')).toBeFalsy()
        expect(items.at(2).prop('selected')).toBeFalsy()
        expect(items.at(3).prop('disabled')).toBeTruthy()
        expect(items.at(4).prop('selected')).toBeFalsy()
        expect(items.at(5).prop('selected')).toBeTruthy()
        expect(items.at(6).prop('disabled')).toBeTruthy()
        expect(items.at(7).prop('disabled')).toBeTruthy()
      })
    })
  })
})
