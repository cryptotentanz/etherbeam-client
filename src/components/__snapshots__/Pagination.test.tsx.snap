// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Default should match snapshot 1`] = `
<nav>
  <ul
    class="my-list-h my-list-center "
  >
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="First page"
      >
        <span
          class="my-icon  "
        >
          <span
            class=" "
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-double-left fa-w-14 "
              data-icon="angle-double-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M223.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L319.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L393.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34zm-192 34l136 136c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9L127.9 256l96.4-96.4c9.4-9.4 9.4-24.6 0-33.9L201.7 103c-9.4-9.4-24.6-9.4-33.9 0l-136 136c-9.5 9.4-9.5 24.6-.1 34z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Previous page"
      >
        <span
          class="my-icon  "
        >
          <span
            class=" "
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-left fa-w-8 "
              data-icon="angle-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Page 1"
      >
        <span
          class=""
        >
          1
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Page 2"
      >
        <span
          class=""
        >
          2
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Page 3"
      >
        <span
          class=""
        >
          3
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Page 4"
      >
        <span
          class=""
        >
          4
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable my-disabled-fg my-active"
        disabled=""
        title="Page 5"
      >
        <span
          class=""
        >
          5
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Page 6"
      >
        <span
          class=""
        >
          6
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Next page"
      >
        <span
          class="my-icon  "
        >
          <span
            class=" "
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-right fa-w-8 "
              data-icon="angle-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </button>
    </li>
    <li
      class="my-listitem "
    >
      <button
        class="my-button-link my-secondary-fg my-clickable"
        title="Last page"
      >
        <span
          class="my-icon  "
        >
          <span
            class=" "
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-double-right fa-w-14 "
              data-icon="angle-double-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
      </button>
    </li>
  </ul>
</nav>
`;
